mempool/cat/pool.go

```diff
@@ -0,0 +1,786 @@
+		store:            newStore(),
+		broadcastCh:      make(chan types.TxKey, 1),
+		txsToBeBroadcast: make(map[types.TxKey]struct{}),
+	}
+
+	for _, opt := range options {
+		opt(txmp)
+	}
+
+	return txmp
+}
+
+// WithPreCheck sets a filter for the mempool to reject a transaction if f(tx)
+// returns an error. This is executed before CheckTx. It only applies to the
+// first created block. After that, Update() overwrites the existing value.
+func WithPreCheck(f mempool.PreCheckFunc) TxPoolOption {
+	return func(txmp *TxPool) { txmp.preCheckFn = f }
+}
+
+// WithPostCheck sets a filter for the mempool to reject a transaction if
+// f(tx, resp) returns an error. This is executed after CheckTx. It only applies
+// to the first created block. After that, Update overwrites the existing value.
+func WithPostCheck(f mempool.PostCheckFunc) TxPoolOption {
+	return func(txmp *TxPool) { txmp.postCheckFn = f }
+}
+
+// WithMetrics sets the mempool's metrics collector.
+func WithMetrics(metrics *mempool.Metrics) TxPoolOption {
+	return func(txmp *TxPool) { txmp.metrics = metrics }
+}
+
+// Lock is a noop as ABCI calls are serialized
+func (txmp *TxPool) Lock() {}
+
+// Unlock is a noop as ABCI calls are serialized
+func (txmp *TxPool) Unlock() {}
+
+// Size returns the number of valid transactions in the mempool. It is
+// thread-safe.
+func (txmp *TxPool) Size() int { return txmp.store.size() }
+
+// SizeBytes return the total sum in bytes of all the valid transactions in the
+// mempool. It is thread-safe.
+func (txmp *TxPool) SizeBytes() int64 { return txmp.store.totalBytes() }
+
+// FlushAppConn executes FlushSync on the mempool's proxyAppConn.
+//
+// The caller must hold an exclusive mempool lock (by calling txmp.Lock) before
+// calling FlushAppConn.
+func (txmp *TxPool) FlushAppConn() error {
+	return txmp.proxyAppConn.FlushSync()
+}
+
+// EnableTxsAvailable enables the mempool to trigger events when transactions
+// are available on a block by block basis.
+func (txmp *TxPool) EnableTxsAvailable() {
+	txmp.txsAvailable = make(chan struct{}, 1)
+}
+
+// TxsAvailable returns a channel which fires once for every height, and only
+// when transactions are available in the mempool. It is thread-safe.
+func (txmp *TxPool) TxsAvailable() <-chan struct{} { return txmp.txsAvailable }
+
+func (txmp *TxPool) Height() int64 {
+	txmp.updateMtx.Lock()
+	defer txmp.updateMtx.Unlock()
+	return txmp.height
+}
+
+func (txmp *TxPool) Has(txKey types.TxKey) bool {
+	return txmp.store.has(txKey)
+}
+
+func (txmp *TxPool) Get(txKey types.TxKey) (types.Tx, bool) {
+	wtx := txmp.store.get(txKey)
+	if wtx != nil {
+		return wtx.tx, true
+	}
+	return types.Tx{}, false
+}
+
+func (txmp *TxPool) IsRejectedTx(txKey types.TxKey) bool {
+	return txmp.rejectedTxCache.Has(txKey)
+}
+
+func (txmp *TxPool) WasRecentlyEvicted(txKey types.TxKey) bool {
+	return txmp.evictedTxs.Has(txKey)
+}
+
+func (txmp *TxPool) CanFitEvictedTx(txKey types.TxKey) bool {
```

Comment:
A description for this method would be helpful specially the part that the transaction is retrieved from the `evictedTxs` and then inserted into the `store` (it may be confused with the case that we want to fit the `txKey` in the `evictedTxs`).

Reply:
I have decided to remove the eviction cache and logic